Add PC-AiR

Will-Tyler opened the pull request:
### Description

In this pull request, I implement the PC-AiR method as described in this [paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4836868/). This includes the PC-AiR algorithm for partitioning samples into an unrelated set and a related set. The method, `pc_air` returns a block matrix with the principle components as the columns of the block matrix.

### Testing

I add unit tests.

I manually inspected the documentation to confirm that it matches my expectations.

konradjk commented:
This is great! From a downstream utility perspective, should this return BlockMatrix, or a hail table keyed by sample (more like what hl.hwe_normalized_pca returns)? I feel like most analytics would then be plotting PCs, or using them in models, but I might be missing an application that needs BlockMatrix

Will-Tyler commented on hail/python/hail/methods/relatedness/pc_air.py:
```diff
@@ -0,0 +1,278 @@
+
+    unrelated = pairs.aggregate_cols(hl.agg.collect_as_set(pairs.col_key))
+    related = set()
+    samples = pairs.cols()
```
I am starting to think that this method (`_partition_samples`) would be better if it localized the samples table and performed the algorithm locally. This method's current implementation is rather slow—it takes 63 seconds to run the unit test I wrote for this method on my machine. Even for studies with millions of samples, it seems reasonable to assume this data could fit on a single machine.

Will-Tyler replied:
I got the execution time of the unit test down to 37 seconds in commit 4c570be by avoiding set operations. I then made the implementation more Hail-idiomatic by using table joins instead of manually filtering and selecting the keys (commit e89f3cf). This brought the execution time up to 40 seconds, but I prefer this approach because of its simplicity.

Still wondering if `_partition_samples` should just localize the samples though...

Will-Tyler commented:
I think I can simplify the computation of the principal components. First let me summarize how the paper presents PC-AiR.

### Original Presentation

Suppose that there are $n_u$ individuals in the unrelated set and $p$ SNPs. Let $X$ be the $n_u \times p$ standardized genotype matrix for the unrelated individuals.

The paper starts by computing

$$\Psi \coloneqq \frac{1}{p}XX^T.$$

Next the paper eigendecomposes $\Psi$:

$$\Psi = U \Sigma^2 U^T,$$

where $U$ are the eigenvectors and $\Sigma^2$ is the diagonal matrix with the eigenvalues along the diagonal.

Then the paper computes a $p \times n$ matrix $W$ called the **SNP weight matrix**:

$$W \coloneqq X^T U.$$

Suppose that there are $n_r$ individuals in the related set and let $Y$ be the $n_r \times p$ standardized genotype matrix for the related individuals. The paper computes the principal components associated with the related samples with

$$ \frac{1}{p} Y W (\Sigma^2)^{-1}.$$

### Simplifications

The first simplification that I noticed is that we can do away with the $\frac{1}{p}$ terms. Because $\Psi$ is scaled by $p^{-1}$, the inverse of the eigenvalues, $(\Sigma^2)^{-1}$ is scaled by $p$, which cancels out the $1/p$ term in the calculation of the principal components for the related individuals. From here on, let us redefine $\Sigma^2$ as the diagonal matrix containing the eigenvalues of $XX^T$ (not $\frac{1}{p} XX^T$).

Next, by examining the relationship between singular value decomposition (SVD) and eigendecomposition ([Wikipedia link](https://en.m.wikipedia.org/wiki/Singular_value_decomposition#Relation_to_eigenvalue_decomposition)), I realized that it is not necessary to compute $\Psi$. Instead, we can get $U$ and $\Sigma$ from the SVD of $X$:

$$X = U\Sigma V^T,$$

where $V$ is a $p \times p$ basis of the new PCA coordinate space.

Then while investigating the meaning of $W$, I realized that $W = X^T U = V \Sigma^T U^T U = V \Sigma^T$.

Taking these simplifications into account, I realized that the paper is, in essence, computing $Y V$ to get the predicted scores associated with the related individuals. (Technically, I think that the paper is computing $Y V \Sigma^{-1}$. I am not sure why they scale the columns by the reciprocal of the eigenvalues here.)

### Simplified Approach

As I understand it, $V$ is a change-of-basis matrix from the original coordinates to the PCA scores. Letting $G$ denote the full standardized genotype matrix of all the individuals, I think we can just return $GV$, where $V$ is defined by the SVD of $X$:

$$X = U\Sigma V^T.$$

patrick-schultz commented:
Hi Will, just want to let you know I started reviewing this yesterday but didn't have to to finish. Overvall this is great work! I'm planning on reviewing the PC-AiR paper over the weekend, and I'll look over your summary of the math (thanks for that!). I'm aiming to have a review ready Monday.

Will-Tyler commented:
Thanks for the update. Let me know if there is anything I can do to help with the review.

patrick-schultz reviewed (changes requested):
This is a really great start!

Besides the comments on speeding up the partitioning step, my biggest issue is with the SVD step. For this to scale to large sample sizes, we can neither localize the entire matrix, nor compute the full SVD. The paper only discusses computing a full eigendecomposition, but in practice this is neither feasable nor necessary. In applications only the PCs with the largest eigenvalues are wanted, and it is possible to compute those much more efficiently than the full SVD.

To avoid both of these issues, I propose you add an optional argument to `_hwe_normalized_blanczos` which specifies a boolean field marking a set of mutually unrelated samples, as inferred by the partitioning algorithm. (This is an experimental PCA implementation which will soon become our default algorithm.) You would need to add this argument to `_blanczos_pca` and `_make_tsm_from_call` also, but I believe that would be it. You would need to modify the normalization in `_make_tsm_from_call` to compute allele frequencies from only the unrelated set, and modify `_blanczos_pca` to compute the svd using only the unrelated set. Then the PC scores of the related set can be computed by multiplying the normalized genotype matrix of the related set with the PC loadings matrix, to project onto the PC subspace.

This is getting into some of the more experimental parts of hail, so please reach out if you need guidance. And if this looks like more than you have time to do, we can try to find a closer target to get this merged in some form, for someone to pick up in the future. I wouldn't want to lose all the progress you've made!

Will-Tyler commented on hail/python/hail/methods/relatedness/pc_air.py:
```diff
@@ -0,0 +1,190 @@
+    standardized_genotypes = hwe_normalize(genotypes)
+    standardized_genotypes = BlockMatrix.from_entry_expr(standardized_genotypes).T
+    scores = standardized_genotypes @ _loadings
+    return eigenvalues, scores, loadings
```
I want to make sure I got the essence of this method correct here. With this approach, I do not modify any of the existing methods.

If this is essentially correct, then I think all that's left to do is determine the best return type to use and perhaps avoid the redundant standardization. (Both `_hwe_normalized_blanczos` and `hwe_normalize` standardize the genotypes.)

For the return types, should I match the format `_hwe_normalized_blanczos` uses, returning the scores and the loadings as tables with one row field keyed by the sample key? (I think this is what @konradjk suggested as well.)

Will-Tyler reviewed (commented):
I think `_partition_samples` is ready for a new look, and I had some questions about PCA part.

patrick-schultz commented:
Re. the questions about the PCA step, I think you'll be beter off modifying `_hwe_normalized_blanczos`. For one thing, this ensures that PC-AiR always returns results in the same form as normal PCA.

More importantly, `_hwe_normalized_blanczos` performs the SVD using a "tall-skinny matrix" representation, which is just a table of matrices (2d ndarrays). This is more efficient than using block matrices for several reasons that aren't directly relevant here. The result of the SVD is computed as local numpy ndarrays. Given these forms of the data, projecting the related sampled onto the computed PCs should be straightforward and efficient. But once everything is converted to tables and matrixtables, it's much harder and does a lot of redundant work.

Let me know if you want to schedule a time to walk through the PCA internals and where you can plug in to them.

Will-Tyler commented on hail/python/test/hail/methods/relatedness/test_pc_air.py:
```diff
@@ -0,0 +1,72 @@
+        self.assertEqual(expected_related, related)
+        self.assertEqual(expected_unrelated, unrelated)
+
+    def test_pc_air(self):
```
I can add more testing—I wanted to make sure that the code is on the right track first though.

patrick-schultz commented on hail/python/test/hail/methods/relatedness/test_pc_air.py:
```diff
@@ -0,0 +1,72 @@
+        table = _partition_samples(mt.GT, 0.05)
+        unrelated = set(table.filter(table.is_in_unrelated).s.collect())
+        related = set(table.filter(~table.is_in_unrelated).s.collect())
+        expected_related = {
```
How did you construct the expected result?

Will-Tyler replied:
I created this by running the code and copying the output, so this unit test serves as a regression test. I will see if I can improve on this.

patrick-schultz reviewed (changes requested):
Looking great! Let me know when you're happy with the tests.

Will-Tyler commented:
For the partitioning algorithm, I updated the test to confirm that all the individuals in the unrelated set are mutually unrelated.

For PC-AiR, I updated the test to compare the loadings to PCA on just the unrelated individuals. The loadings are NumPy close. The scores are slightly different though because they are calculated differently. When there are related individuals, the scores are calculated by multiplying the loadings and the standardized genotypes. When there are no related individuals, the scores are calculated by multiplying the columns of the appropriate singular matrix with the eigenvalues. So for the scores, I just add a regression test. (In my testing, I observed that most of the scores were less than 1% different. However, there were a few differences that were larger around 20% or 30%. Not sure if this is a cause for concern because the calculation approaches are different and the SVD is approximate.)

patrick-schultz commented:
`test_pc_air` and `test_partition_samples` are both failing, though the latter is only failing on the local backend for some reason.
